master/internal/resourcemanagers/fitting_methods.go

```diff
@@ -12,20 +12,62 @@ func labelSatisfied(req *AllocateRequest, agent *agentState) bool {
 	return req.Label == agent.label
 }
 
+func maxZeroSlotContainersSatisfied(req *AllocateRequest, agent *agentState) bool {
+	if req.SlotsNeeded == 0 {
+		if agent.maxZeroSlotContainers == nil {
+			return true
+		}
+		if *agent.maxZeroSlotContainers == 0 {
+			return false
+		}
+		return agent.numZeroSlotContainers() < *agent.maxZeroSlotContainers
+	}
+	return true
+}
+
+func agentIdleSatisfied(_ *AllocateRequest, agent *agentState) bool {
+	return agent.idle()
+}
+
 // Soft Constraints
 
 // BestFit returns a float affinity score between 0 and 1 for the affinity between the task and
 // the agent. This method attempts to allocate tasks to the agent that is both most utilized and
 // offers the fewest slots. This method should be used when the cluster is dominated by multi-slot
 // applications.
-func BestFit(_ *AllocateRequest, agent *agentState) float64 {
+func BestFit(req *AllocateRequest, agent *agentState) float64 {
+	if req.SlotsNeeded == 0 {
+		if agent.maxZeroSlotContainers == nil {
```

Comment:
@shiyuann @armandmcqueen what do you think about the approach where if `maxZeroSlotContainers` is not specified we fall back on the old method instead for both this and `WorstFit`?

Reply:
By old method I mean using the number of slots used rather than number of zero slot containers running.

Reply:
It would be bad because for CPU resource pools tasks would be scheduled only according to hash distances between agent id and task id

Reply:
@shiyuann and I talked offline and our current thought is to use the old approach when an agent has slots and when the agent doesn't have slots to use the new approach.